src/System.Net.Security/src/System/Net/SecureProtocols/_SslStream.cs

```diff
@@ -0,0 +1,709 @@
+            }
+
+            EnsureInternalBufferSize(SecureChannel.ReadHeaderSize, readBytes);
+
+            if (asyncRequest != null) //Async
+            {
+                asyncRequest.SetNextRequest(InternalBuffer, SecureChannel.ReadHeaderSize, readBytes, s_readFrameCallback);
+
+                _Reader.AsyncReadPacket(asyncRequest);
+
+                if (!asyncRequest.MustCompleteSynchronously)
+                {
+                    return 0;
+                }
+                readBytes = asyncRequest.Result;
+            }
+            else //Sync
+            {
+                readBytes = _Reader.ReadPacket(InternalBuffer, SecureChannel.ReadHeaderSize, readBytes);
+            }
+            return ProcessFrameBody(readBytes, buffer, offset, count, asyncRequest);
+        }
+
+        //
+        // readBytes == SSL Data Payload size on input or 0 on EOF
+        //
+        private int ProcessFrameBody(int readBytes, byte[] buffer, int offset, int count, AsyncProtocolRequest asyncRequest)
+        {
+            if (readBytes == 0)
+            {
+                // EOF
+                throw new IOException(SR.net_io_eof);
+            }
+
+            // Set readBytes to total number of received bytes.
+            readBytes += SecureChannel.ReadHeaderSize;
+
+            // Decrypt into internal buffer, change "readBytes" to count now _Decrypted Bytes_.
+            int data_offset = 0;
+
+            Interop.SecurityStatus errorCode = _SslState.DecryptData(InternalBuffer, ref data_offset, ref readBytes);
+
+            if (errorCode != Interop.SecurityStatus.OK)
+            {
+                byte[] extraBuffer = null;
+                if (readBytes != 0)
+                {
+                    extraBuffer = new byte[readBytes];
+                    Buffer.BlockCopy(InternalBuffer, data_offset, extraBuffer, 0, readBytes);
+                }
+                // Reset internal buffer count.
+                DecrementInternalBufferCount(InternalBufferCount);
+                return ProcessReadErrorCode(errorCode, buffer, offset, count, asyncRequest, extraBuffer);
+            }
+
+
+            if (readBytes == 0 && count != 0)
+            {
+                // Read again since remote side has sent encrypted 0 bytes.
+                DecrementInternalBufferCount(InternalBufferCount);
+                return -1;
+            }
+
+            // Decrypted data start from "data_offset" offset, the total count can be shrunk after decryption.
+            EnsureInternalBufferSize(0, data_offset + readBytes);
+            DecrementInternalBufferCount(data_offset);
+
+            if (readBytes > count)
+            {
+                readBytes = count;
+            }
+            Buffer.BlockCopy(InternalBuffer, InternalOffset, buffer, offset, readBytes);
+
+            // This will adjust both the remaining internal buffer count and the offset.
+            DecrementInternalBufferCount(readBytes);
```

Comment:
I think the thing that confuses me is that it's named "BufferCount" instead of "BytesRemaining".  And the method to me feels like it's "Seek{Forward}" or "SkipBytes", not "DecrementCount".